File: BackupUpdater.java
import java.io.File;
import java.util.Arrays;

public class BackupUpdater {
    public static void main(String args[]) {
        String folder_path = "myFolder";
        String backup_path = "myFolder-backup";

        File folder = new File(folder_path);
        File backup = new File(backup_path);

        boolean folder_valid, backup_valid;

        folder_valid = checkValidity(folder, "folder");
        backup_valid = checkValidity(backup, "backup");

        if(folder_valid && backup_valid) {
            compareFolders(folder, backup, folder_path, backup_path);
        }
    } // end of main

    static boolean checkValidity(File f, String name_type) {
        if((!f.exists())) {
            System.out.println("The " + name_type + " path you have provided does not exist.");
            return false;
        } else {
            if(f.isFile()) {
                System.out.println("The " + name_type + " path you have provided is not a directory.");
                return false;
            } else {
                System.out.println("Valid!");
                return true;
            }
        } // end of else
    }

    static void compareFolders(File dir_1, File dir_2, String path_1, String path_2) {
        // list folder content
        String[] list_1 = dir_1.list();
        String[] list_2 = dir_2.list();

        

        /* int file_count_1 = 0;
        int file_count_2 = 0;

        for(int i=0; i<list_1.length; i++) {
            file_count_1++;
        }
        for(int i=0; i<list_2.length; i++) {
            file_count_2++;
        } 

        System.out.println(file_count_1);
        System.out.println(file_count_2); */

        if(Arrays.equals(list_1, list_2)) {
            // compare properties of all files
            for(int i=0; i<list_1.length; i++) {
                File f1 = new File(path_1 + "/" + list_1[i]);
                File f2 = new File(path_2 + "/" + list_2[i]);

                if(f1.lastModified() == f2.lastModified()) {
                    continue;
                } else {

                }
            }
        } else {
            System.out.println("updating backup folder...");
        }

        // compare properties of files that are the same
        // if different properties, take file from dir_1 and replace said file in dir_2

        // if dir_1 has file that dir_2 doesnt have, copy said file from dir_1 to dir_2
    }
}
// placeholder file

File: backup-updater.py
import os 
import shutil 
import time

folderPath = "C:/Users/MrRobot/Desktop/portfolio - Copy"
backupPath = "C:/Users/MrRobot/Desktop/New folder"

def exists(path):
    if os.path.exists(path) and os.path.isdir(path): # Check if path exists and is a directory
        print(str(path) + " - is a path and a directory")
        return True
    elif os.path.isfile(path):
        print(str(path) + " - is a file and not a folder.") # Not a directory
        return False
    elif not os.path.exists(path):
        print(str(path) + " - is invalid or does not exist.") # Invalid path given
        return False
    
def checkTypeAndCopy(x, y):
    if os.path.isfile(x):
        if shutil.copy2(x, y):
            return True
        
    elif os.path.isdir(x):
        print(x)
        print(y)
        if shutil.copytree(x, y):
            return True
    
def checkModDateAndSize(fileToCopy, dest, i):
    ftcPath = os.path.join(folderPath, fileToCopy[i])
    dPath = os.path.join(backupPath, dest[i])

    ftcModTime = os.path.getmtime(ftcPath)
    dModTime = os.path.getmtime(dPath)

    ftcSize = os.path.getsize(ftcPath)
    dSize = os.path.getsize(dPath)

    if (ftcModTime != dModTime) or (ftcSize != dSize):
        print(f'"{fileToCopy[i]}"' + " is out of date.")
        print("Updating " + f'"{fileToCopy[i]}"' + "...")
        if checkTypeAndCopy(ftcPath, dPath):
            print(f'"{fileToCopy[i]}"' + " is now up to date.")
            print("")

def backupNewFile(fileToBackup):
    ftbPath = os.path.join(folderPath, fileToBackup)
    print(f'"{fileToBackup}"' + " hasn't been backed up.")
    print("Backing up " + f'"{fileToBackup}"' + "...")
    if checkTypeAndCopy(ftbPath, backupPath + "/" + fileToBackup): # change path
        print("Success!")
        print("")

# END OF METHODS


while True:
    backupDone = False
    if not backupDone:
        if exists(folderPath) and exists(backupPath): # If both directories exist/are valid
            folderFiles = os.listdir(folderPath) # List the content of both directories
            backupFiles = os.listdir(backupPath)
            fileCount = len(folderFiles) # Get file count
            
            if folderFiles == backupFiles: # Check if both directories have the same content
                for i in range(fileCount):
                    checkModDateAndSize(folderFiles, backupFiles, i)
            # If directory content is not the same
            else: 
                for i in folderFiles:
                    if i not in backupFiles: # Look for files missing in the backup folder & back them up
                        backupNewFile(i)
                        
    backupDone = True
    if backupDone:
        print("")
        print("Backup folder is up to date!!")
    time.sleep(10)